# Importing necessary libraries

In [0]:
#requets library helps us to access required content through urls
import requests

#pandas library helps us in converting data in variious formats to dataframes which are easier to work with
import pandas as pd

#Folium library helps in working with maps
import folium

#for converting an address into latitude and longitude values
from geopy.geocoders import Nominatim

import numpy as np

import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

# Retrieve data stored in google drive into a pandas datafame.
#### One can find the original file in the link provided in the report

In [0]:
path='/content/drive/My Drive/crime.csv'
crime=pd.read_csv(path, sep=',', engine='python')
crime.head()

INCIDENT_NUMBER  OFFENSE_CODE  ...       Long                     Location
0      I182080058          2403  ... -71.121186  (42.26260773, -71.12118637)
1      I182080053          3201  ... -71.135311  (42.35211146, -71.13531147)
2      I182080052          2647  ... -71.076930  (42.30812619, -71.07692974)
3      I182080051           413  ... -71.059648  (42.35945371, -71.05964817)
4      I182080050          3122  ... -71.024663  (42.37525782, -71.02466343)

[5 rows x 17 columns]

# Analyzing Data

In [0]:
# To obtain total no.of streets with repective no.of crime entries
crime['STREET'].nunique()


4684

 There are totally **4684** rows. let's deal with the high crime rate areas so the end user can avoid them.

In [0]:
high_crime=crime['STREET'].value_counts().to_frame()

#To extract top 30 crime rated areas
high_crime=high_crime.head(30)
high_crime.reset_index(inplace=True)
high_crime.columns=['STREET','CRIME_COUNT']

#collect the co-ordinates of these 30 locations
high_crime=high_crime.merge(crime,on=['STREET'])
high_crime=high_crime[['STREET','CRIME_COUNT','Lat','Long']]

#When we merged data all the thousands of rows will be merged with it.This increases API calls
#So we shall remove these duplicates
high_crime=high_crime.drop_duplicates('STREET')
high_crime.sort_values(by='STREET',inplace=True,ascending=True)
high_crime.reset_index(inplace=True)
high_crime.head()

index               STREET  CRIME_COUNT        Lat       Long
0  78683             ADAMS ST         2104  42.305378 -71.061514
1  93191            ALBANY ST         1466  42.330769 -71.077665
2  94657          ALLSTATE RD         1332  42.326966 -71.061986
3  89858  AMERICAN LEGION HWY         1672  42.297950 -71.087744
4  80787            BEACON ST         2016  42.355052 -71.073907

In [0]:
#lets keep only crime data that is relevant for analysis
streets=high_crime['STREET']
crime=crime[crime['STREET'].isin(streets)]
crime.reset_index(drop=True,inplace=True)
crime=crime[['STREET','OFFENSE_CODE_GROUP','Lat','Long']]
crime.head()

STREET               OFFENSE_CODE_GROUP        Lat       Long
0      CAMBRIDGE ST               Aggravated Assault  42.359454 -71.059648
1    DORCHESTER AVE                        Vandalism  42.299197 -71.060470
2        TREMONT ST                  Verbal Disputes  42.333807 -71.103778
3  COMMONWEALTH AVE                            Towed  42.348866 -71.089363
4       COLUMBIA RD  Motor Vehicle Accident Response  42.330015 -71.038513

k-means cluster algorithm works only with numeric data. But the crime category is a string. Using get_dummies method in pandas we can create a separate column for each category that depicts its row value with 0 or 1. This is called one-hot encoding

In [0]:
crime_k=pd.get_dummies(crime[['OFFENSE_CODE_GROUP']],prefix="",prefix_sep="")
crime_k['STREET']=crime['STREET']

#move street column to starting
cols=[crime_k.columns[-1]]+list(crime_k.columns[:-1])
crime_k=crime_k[cols]
crime_k.head()

STREET  Aggravated Assault  ...  Violations  Warrant Arrests
0      CAMBRIDGE ST                   1  ...           0                0
1    DORCHESTER AVE                   0  ...           0                0
2        TREMONT ST                   0  ...           0                0
3  COMMONWEALTH AVE                   0  ...           0                0
4       COLUMBIA RD                   0  ...           0                0

[5 rows x 67 columns]

let's find mean of each category group of crimes presence to obtain a normalized value and it decreases no.of data rows

In [0]:
crime_grouped=crime_k.groupby(by=['STREET']).mean().reset_index()
print(crime_grouped.shape)
crime_grouped.head()

(30, 67)


STREET  Aggravated Assault  ...  Violations  Warrant Arrests
0             ADAMS ST            0.017110  ...    0.032319         0.028992
1            ALBANY ST            0.029332  ...    0.015689         0.085266
2          ALLSTATE RD            0.014264  ...    0.012012         0.072072
3  AMERICAN LEGION HWY            0.016746  ...    0.048445         0.013158
4            BEACON ST            0.011905  ...    0.013889         0.012401

[5 rows x 67 columns]

In [0]:
#CLIENT_ID = 'HAVSYF2EXYE3ZZYGUNAED0X2TA0H3PN4HHLAARFE0SEW41UZ' # your Foursquare ID
#CLIENT_SECRET = 'YLFWFYECMOVPL0JNPVWOR4YR2PETGPBWOBLX4JF1C0EZER2Q' # your Foursquare Secret
CLIENT_ID='A0ZWWRRIUV4Y2BO2W2E4QHTOV3KYV3ALZA0UTSJ5B1F3QNYA'
CLIENT_SECRET='SAESVBMDTVTSOX2YNO5VOFGT1VJVTPOA2B0KWEQY555NJNPS'
VERSION = '20200524'

In [0]:
#function to obtain details of venues in each neighborhood using foursquare
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    LIMIT=50
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [0]:
boston_venues=getNearbyVenues(names=high_crime['STREET'],
                               latitudes=high_crime['Lat'],
                               longitudes=high_crime['Long'])
boston_venues.head()

ADAMS ST
ALBANY ST
ALLSTATE RD
AMERICAN LEGION HWY
BEACON ST
BLUE HILL AVE
BORDER ST
BOWDOIN ST
BOYLSTON ST
CAMBRIDGE ST
CENTRE ST
COLUMBIA RD
COLUMBUS AVE
COMMONWEALTH AVE
DORCHESTER AVE
DUDLEY ST
GENEVA AVE
HANCOCK ST
HARRISON AVE
HUNTINGTON AVE
HYDE PARK AVE
MASSACHUSETTS AVE
NEW SUDBURY ST
NEWBURY ST
RIVER ST
SUMMER ST
TREMONT ST
W BROADWAY
WARREN ST
WASHINGTON ST


Neighborhood  Neighborhood Latitude  ...  Venue Longitude           Venue Category
0     ADAMS ST              42.305378  ...       -71.059102    Vietnamese Restaurant
1     ADAMS ST              42.305378  ...       -71.059008                     Café
2     ADAMS ST              42.305378  ...       -71.060053    Vietnamese Restaurant
3     ADAMS ST              42.305378  ...       -71.058305  New American Restaurant
4     ADAMS ST              42.305378  ...       -71.058642                 Dive Bar

[5 rows x 7 columns]

In [0]:
cols=['STREET','STREET Latitude', 
                  'STREET Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
boston_venues.columns=cols
boston_venues.head()

STREET  STREET Latitude  ...  Venue Longitude           Venue Category
0  ADAMS ST        42.305378  ...       -71.059102    Vietnamese Restaurant
1  ADAMS ST        42.305378  ...       -71.059008                     Café
2  ADAMS ST        42.305378  ...       -71.060053    Vietnamese Restaurant
3  ADAMS ST        42.305378  ...       -71.058305  New American Restaurant
4  ADAMS ST        42.305378  ...       -71.058642                 Dive Bar

[5 rows x 7 columns]

In [0]:
# one hot encoding of venues
boston_onehot = pd.get_dummies(boston_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
boston_onehot['STREET'] = boston_venues['STREET'] 

# move neighborhood column to the first column
fixed_columns = [boston_onehot.columns[-1]] + list(boston_onehot.columns[:-1])
boston_onehot = boston_onehot[fixed_columns]

boston_onehot.head()

STREET  ATM  Accessories Store  ...  Wine Shop  Women's Store  Yoga Studio
0  ADAMS ST    0                  0  ...          0              0            0
1  ADAMS ST    0                  0  ...          0              0            0
2  ADAMS ST    0                  0  ...          0              0            0
3  ADAMS ST    0                  0  ...          0              0            0
4  ADAMS ST    0                  0  ...          0              0            0

[5 rows x 177 columns]

In [0]:
#find mean for normalized value
boston_grouped = boston_onehot.groupby('STREET').mean().reset_index()
boston_grouped.head()

STREET  ATM  ...  Women's Store  Yoga Studio
0             ADAMS ST  0.0  ...           0.00          0.0
1            ALBANY ST  0.0  ...           0.00          0.0
2          ALLSTATE RD  0.0  ...           0.00          0.0
3  AMERICAN LEGION HWY  0.0  ...           0.00          0.0
4            BEACON ST  0.0  ...           0.02          0.0

[5 rows x 177 columns]

Merge dummy variables of crimes and venue data for clustering

In [0]:
merged=crime_grouped.merge(boston_grouped,on='STREET')
merged.head()

STREET  Aggravated Assault  ...  Women's Store  Yoga Studio
0             ADAMS ST            0.017110  ...           0.00          0.0
1            ALBANY ST            0.029332  ...           0.00          0.0
2          ALLSTATE RD            0.014264  ...           0.00          0.0
3  AMERICAN LEGION HWY            0.016746  ...           0.00          0.0
4            BEACON ST            0.011905  ...           0.02          0.0

[5 rows x 243 columns]

function to return most common values

In [0]:
def return_most_common(row, num_top):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

To form a dataframe that contains top 10 venues of respective areas

In [0]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['STREET']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
venues_sorted = pd.DataFrame(columns=columns)
venues_sorted['STREET'] = boston_grouped['STREET']

for ind in np.arange(boston_grouped.shape[0]):
   venues_sorted.iloc[ind, 1:] = return_most_common(boston_grouped.iloc[ind, :], num_top_venues)

venues_sorted.head()

STREET  ...   10th Most Common Venue
0             ADAMS ST  ...  New American Restaurant
1            ALBANY ST  ...         Sushi Restaurant
2          ALLSTATE RD  ...               Shoe Store
3  AMERICAN LEGION HWY  ...                  Dog Run
4            BEACON ST  ...                      Gym

[5 rows x 11 columns]

To form a dataframe with top 10 common crimes in respective areas

In [0]:
num_top_crimes = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['STREET']
for ind in np.arange(num_top_crimes):
    try:
        columns.append('{}{} Common crime'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Common crime'.format(ind+1))

# create a new dataframe
crimes_sorted = pd.DataFrame(columns=columns)
crimes_sorted['STREET'] = crime_grouped['STREET']

for ind in np.arange(crime_grouped.shape[0]):
   crimes_sorted.iloc[ind, 1:] = return_most_common(crime_grouped.iloc[ind, :], num_top_crimes)

crimes_sorted.head()

STREET  ... 10th Common crime
0             ADAMS ST  ...        Violations
1            ALBANY ST  ...     Property Lost
2          ALLSTATE RD  ...    Simple Assault
3  AMERICAN LEGION HWY  ...         Vandalism
4            BEACON ST  ...             Other

[5 rows x 11 columns]

# Clustering of locations with k-means clustering algorithm

In [0]:
# set number of clusters
kclusters = 5

boston_grouped_clustering = merged.drop('STREET', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(boston_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

# add clustering labels
merged.insert(0, 'Cluster Labels', kmeans.labels_)

In [0]:
merged[['Lat']]=high_crime[['Lat']]
merged[['Long']]=high_crime[['Long']]
merged=merged[['STREET','Lat','Long','Cluster Labels']]
merged=merged.merge(crimes_sorted,on=['STREET'])
merged=merged.merge(venues_sorted,on=['STREET'])

merged.shape

(30, 24)

In [0]:
merged.head()

STREET  ...   10th Most Common Venue
0             ADAMS ST  ...  New American Restaurant
1            ALBANY ST  ...         Sushi Restaurant
2          ALLSTATE RD  ...               Shoe Store
3  AMERICAN LEGION HWY  ...                  Dog Run
4            BEACON ST  ...                      Gym

[5 rows x 24 columns]

In [0]:
address = 'Boston'

geolocator = Nominatim(user_agent="boston_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude


In [0]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
print(ys)
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(merged['Lat'], merged['Long'], merged['STREET'], merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

[array([0, 1, 2, 3, 4]), array([ 1,  3,  7, 13, 21]), array([ 2,  7, 20, 41, 70]), array([  3,  13,  41,  87, 151]), array([  4,  21,  70, 151, 264])]


# Examining clusters

In [0]:
merged.loc[merged['Cluster Labels']==0,merged.columns[[0]+list(range(4,merged.shape[1]))]]

STREET  ...   10th Most Common Venue
0            ADAMS ST  ...  New American Restaurant
1           ALBANY ST  ...         Sushi Restaurant
6           BORDER ST  ...             Liquor Store
10          CENTRE ST  ...            Grocery Store
14     DORCHESTER AVE  ...           Baseball Field
15          DUDLEY ST  ...        Mobile Phone Shop
16         GENEVA AVE  ...            Metro Station
17         HANCOCK ST  ...                     Lake
18       HARRISON AVE  ...     Gym / Fitness Center
20      HYDE PARK AVE  ...           Discount Store
21  MASSACHUSETTS AVE  ...            Shopping Mall
22     NEW SUDBURY ST  ...       Salon / Barbershop
24           RIVER ST  ...        Convenience Store
26         TREMONT ST  ...       Italian Restaurant
27         W BROADWAY  ...           Rental Service
28          WARREN ST  ...        Electronics Store
29      WASHINGTON ST  ...                     Café

[17 rows x 21 columns]

From above dataframe we can observe that the top crimes in this cluster of areas is motor vehicle accident,larceny,drug violation

In [0]:
merged.loc[merged['Cluster Labels']==1,merged.columns[[0]+list(range(4,merged.shape[1]))]]

STREET  ... 10th Most Common Venue
2        ALLSTATE RD  ...             Shoe Store
4          BEACON ST  ...                    Gym
8        BOYLSTON ST  ...                Dog Run
9       CAMBRIDGE ST  ...    American Restaurant
11       COLUMBIA RD  ...      Electronics Store
12      COLUMBUS AVE  ...                    Gym
13  COMMONWEALTH AVE  ...       Ramen Restaurant
19    HUNTINGTON AVE  ...            Coffee Shop
23        NEWBURY ST  ...                  Hotel
25         SUMMER ST  ...           Burger Joint

[10 rows x 21 columns]

In [0]:
merged.loc[merged['Cluster Labels']==2,merged.columns[[0]+list(range(4,merged.shape[1]))]]

STREET  ... 10th Most Common Venue
3  AMERICAN LEGION HWY  ...                Dog Run

[1 rows x 21 columns]

In [0]:
merged.loc[merged['Cluster Labels']==3,merged.columns[[0]+list(range(4,merged.shape[1]))]]

STREET  ... 10th Most Common Venue
7  BOWDOIN ST  ...                Dog Run

[1 rows x 21 columns]

In [0]:
merged.loc[merged['Cluster Labels']==4,merged.columns[[0]+list(range(4,merged.shape[1]))]]

STREET  ... 10th Most Common Venue
5  BLUE HILL AVE  ...                Dog Run

[1 rows x 21 columns]